### ISSUE TITLE
Async Cluster: AWS Failover takes a long time to recover

### ISSUE THREAD

**Issue by kamulos** [redis-rs/redis-rs#1005]
I have a MemoryDB Cluster in AWS and am testing how my service handles failover situations. I do this by having my service do regular health checks and then pressing the "Failover primary" Button in the AWS Console. My observation is that it takes 3 Minutes for `redis-rs` to recover to a fully working state.

This is what I observe: AWS provides me with a cluster endpoint, that I provide as single "initial node" to `redis-rs`. My cluster has 4 instances: 2 masters and 2 replicas.

I hope my interpretation here is correct: If I trigger the failover, AWS takes one of the masters away and the associated replica takes over. Then AWS sets up a completely new node and integrates it into the cluster as new replica instance. `redis-rs` seems to not know that there is already a new working master instance. Instead it tries to reconnect to the old master DNS name. This takes a while until the new replica is set up under this DNS name. As soon as `redis-rs` is able to connect to the newly integrated node, it discovers, the failed over master that has been there the whole time.

I am not sure if it makes sense for `redis-rs` to memorize all nodes of the cluster to quickly rediscover new masters. What do you think?

**Comment by kamulos** [comment]
@nihohit I am on it, but I might only really get into it sometime in January.

I updated my copy to the newest master (before it was also pretty recent), and at the moment I am unable to reproduce the issue. Unfortunately I also only have 5 failovers a day in AWS so it is kind of a lengthy process. But it might be, that this issue resolved itself.

I put in quite a bit of debug logging to make things more understandable. I might try to upstream those, because the current trace log messages are not very useful for debugging these kinds of problems.

**Comment by kamulos** [comment]
I got back to it: with release 0.24.0 I have the issue, that the MemoryDB Failover takes minutes, with the current main branch version it is fixed and takes 17 seconds.

The reason is simple: the timeouts of #834 . Without a connection timeout `redis-rs` will hang in `refresh_slots` and prevent the library from updating its list of connections.

Now the fun part: I never configured a timeout. I get a free 1 second timeout because I use the `ClusterClientBuilder`. There in the function `ClusterParams::from(value: BuilderParams)` the default `connection_timeout` is 1s. This in contrast to other places in the code, where the default timeout for both kinds of timeouts is `Duration::MAX`.

So basically the remaining question would be: is this 1s timeout intentional? I think this whole issue is a quite compelling case for enabling some kind of default timeout, maybe even for requests. On the other hand everyone's environment is different and it is difficult to find a default that works well for all of them.

**Comment by jaymell** [comment]
Well, the 1-second timeout was definitely intentional because I asked for it :laughing:. I think some sane value here is particularly important in the clustered code since multiple connections are regularly re-created over time, and nodes may come and go even while the cluster continues to function as a whole. 

I'm not opposed to changing defaults for connection timeouts for the non-clustered clients, but I'm less sure about having default request timeouts.

**Comment by nihohit** [comment]
> with the current main branch version it is fixed and takes 17 seconds.

Hurrah? I think that this deserves a hurrah! 🥳 

>  I think this whole issue is a quite compelling case for enabling some kind of default timeout, maybe even for requests.

I agree with the sentiment, but this might break existing behavior. Let's assume that we decide on a 1 second connection timeout throughout. Now, anyone who has wrapped the MultiplexedConnection connection function with a timeout longer than a second would have their timeout be cut short without notification.
My personal preference is to give the correct defaults to the user, even if it might require them to adjust their code. I'm just not sure how to communicate in this case that the adjustment is required.

**Comment by kamulos** [comment]
As this issue is resolved and there is no obvious way to get to default timeouts, I will close this issue :)
